sklearn/datasets/samples_generator.py

```diff
@@ -754,35 +760,76 @@ def make_blobs(n_samples=100, n_features=2, centers=3, cluster_std=1.0,
     """
     generator = check_random_state(random_state)
 
-    if isinstance(centers, numbers.Integral):
-        centers = generator.uniform(center_box[0], center_box[1],
-                                    size=(centers, n_features))
+    if isinstance(n_samples, numbers.Integral):
+        # n_centers is figured by the centers arg
+        if centers is None:
+            centers = 3
+
+        if isinstance(centers, numbers.Integral):
+            n_centers = centers
+            centers = generator.uniform(center_box[0], center_box[1],
+                                        size=(n_centers, n_features))
+
+        else:
+            centers = check_array(centers)
+            n_features = centers.shape[1]
+            n_centers = centers.shape[0]
+
     else:
-        centers = check_array(centers)
-        n_features = centers.shape[1]
+        # n_centers is figured by the [n_samples] arg
+        n_centers = len(n_samples)
+        if centers is None:
+            # generate centers based on [n_samples]
+            centers = generator.uniform(center_box[0], center_box[1],
+                                        size=(n_centers, n_features))
+        try:
+            assert len(centers) == n_centers
+        except TypeError:
+            raise ValueError("Parameter `centers` should be a"
+                             "list or a np.array.")
+        except AssertionError:
+            raise ValueError("Length of `n_samples` not consistent"
+                             " with number of centers.")
+        else:
+            centers = check_array(centers)
+            n_features = centers.shape[1]
+
+    # stds: if cluster_std is given as list, it must be consistent
+    # with the n_centers
+    if ((isinstance(cluster_std, list) or
```

Comment:
As a general principle, we try to avoid isinstance in python. This will reject a pandas.Series without good reason

Reply:
What should be the best alternative then?
Use a `try except`clause as done above?

Reply:
Yes, that's one common way. Otherwise using hasattr